bin/cliain/src/contracts.rs

```diff
@@ -58,48 +33,47 @@ fn storage_deposit(storage_deposit_limit: Option<u128>) -> Option<Compact<u128>>
     storage_deposit_limit.map(Compact)
 }
 
-pub fn upload_code(
+pub async fn upload_code(
     signed_connection: SignedConnection,
     command: ContractUploadCode,
-) -> anyhow::Result<ContractCodeStoredEvent> {
+) -> anyhow::Result<CodeStored> {
     let ContractUploadCode {
         wasm_path,
         storage_deposit_limit,
     } = command;
 
-    let connection = signed_connection.as_connection();
-
     let wasm = fs::read(wasm_path).expect("WASM artifact not found");
     debug!(target: "contracts", "Found WASM contract code {:?}", wasm);
 
-    let xt = compose_extrinsic!(
-        connection,
-        "Contracts",
-        "upload_code",
-        wasm, // code
-        storage_deposit(storage_deposit_limit)
-    );
-
-    debug!(target: "contracts", "Prepared `upload_code` extrinsic {:?}", xt);
-
-    let _block_hash = send_xt(&connection, xt, Some("upload_code"), XtStatus::InBlock);
-
-    let code_stored_event: ContractCodeStoredEvent = wait_for_event(
-        &connection,
-        ("Contracts", "CodeStored"),
-        |e: ContractCodeStoredEvent| {
-            info!(target : "contracts", "Received CodeStored event {:?}", e);
-            true
-        },
-    )?;
+    let connection = signed_connection.clone();
+    let event_handler = tokio::spawn(async move {
+        connection
+            .wait_for_event(
+                |e: &CodeStored| {
+                    info!(target : "contracts", "Received CodeStored event {:?}", e);
+                    true
+                },
+                BlockStatus::Finalized,
+            )
+            .await
+    });
+
+    let _block_hash = signed_connection
+        .upload_code(
+            wasm,
+            storage_deposit(storage_deposit_limit),
+            TxStatus::InBlock,
+        )
+        .await?;
+    let code_stored_event = event_handler.await?;
```

Comment:
In a separate PR, I think we should use the new [`get_txn_events`](https://github.com/Cardinal-Cryptography/aleph-node/blob/main/aleph-client/src/contract/event.rs#L55) approach here that is more resilient to accidental consumption of someoneelse's event.

Reply:
This applies to almost all methods in this file most probably.

Reply:
This is merged code, not interested in changing any of this.